### ISSUE TITLE
Scrolling helper for all modules

### ISSUE THREAD

**Issue by lasers** [ultrabug/py3status#826]
If modules does not have `on_click` method, should we help them with this function?
```
    def on_click(self, event):
        # Assign buttons
        buttons = (None, 'left', 'middle', 'right', 'up', 'down')
        try:
            button = buttons[event['button']]
        except IndexError:
            return

        # Avoid hitting rate limits on APIs
        if button in ('up', 'down'):
            self.py3.prevent_refresh()
```
This method is targeted at API modules, but could work fine with all modules.

Opt 1) We could add that `on_click` method in only API modules as to ensure that nobody will make massive API calls (accidentally or intentional) due to mouse scrolling, but still allow calls via left/right clicks.

Opt 2) Scrolling helper to avoid mouse scrolling for all modules without `on_click`.

Opt 3) Nothing. Close this.

**Comment by tobes** [comment]
> buttons = (None, 'left', 'middle', 'right', 'up', 'down')

I think we should stick with numbers, as we do not actually know what they refer to. eg could be a multi-touch on a touchpad or some other device

> This method is targeted at API modules,

what do you mean by API module?

the scroll buttons is an interesting idea as they can generate lots of events so maybe we should do something. we can allow user config but default to sensible eg 4, 5 etc or we could do some sort of rate-limiting.  It gets complicated because of how i3bar seems to generate the events.

Anyhow if we do things then we should be doing them in events.py rather than at the module level so they are done universally.

There are other things we should do around clicks

* allow more buttons.  We limit buttons to 1-5 but I can easily generate 6 and 7 on my laptop.  There is no reason to limit this.

* i3pystatus implements double clicks and I think that we should do this too.  It is easy to implement (the i3pystatus module does this)  my suggestion is to add 20 to the button number for double clicks eg double_click is 21 double_middle_click is 22 etc (assuming mouse)

**Comment by lasers** [comment]
>what do you mean by API module?

**Modules with API / JSON / requests.**

`aws_bill`, `bitcoin_prices`, `exchange_rate`, `getjson`, `pingdom`, `etc...`

From `aws_bill`:
>**WARNING: This module generate some costs on the AWS bill.
>Take care about the cache_timeout to limit these fees!**

This uses default `cache_timeout = 3600` with no massive calls due to mouse scrolling safeguard.

Module `github` belongs in this category. I assume we both are using `github` module and there is a good chance we both scrolled this module slightly from time to time.


**Other modules with `command_output` or something similar.**

Other modules such as `dpms` because we shouldn't bother our systems running several `py3.command_output` -> `xset -q` calls in such a short time for no gain.

If you can generate more buttons, then the code should be modified to get all events. This code is a good starter / concept that should work with most comfort mouses.

I'm not feeling this double clicks idea, but I suppose we should play with it first to see how we feels about it versus single clicks before deciding. I feel it might be an annoyance since all modules are mostly simple. Will see how that goes.

Do we know that if `20` is more than enough? Why not `100` to make it obvious? I'm speaking of avoiding mouses like this... https://gizmodo.com/5918345/this-20-button-mouse-puts-your-everything-to-shame

**Comment by tobes** [comment]
With clicks on modules.

If on_click event are set in the config then they will be actioned and the module refreshed.

If a module has an on_click function that will also be called and the module refreshed unless the module requests it not to.

finally if there was no on_click function or one set in the config for the button we refresh the module if the button is 2 as the 'default' action.

So I think this PR is not needed.  However if this is not happening or you see a flaw in the logic then that is an issue that needs to be addressed.  So I think we can close this but maybe I am missing something - or we need to better document this (patches welcome)

modules also pass clicks up to any containers they are in

**Comment by lasers** [comment]
Experiment with `github` module...
```
 diff --git a/py3status/modules/github.py b/py3status/modules/github.py
 index 1c05b15b..505fe375 100644
 --- a/py3status/modules/github.py
 +++ b/py3status/modules/github.py
 @@ -109,6 +109,7 @@ class Py3status:
              auth = None
          try:
              info = self.py3.request(url, timeout=10, auth=auth)
 +            self.py3.log("----- GITHUB REQUEST -----")
          except (self.py3.RequestException):
              return
          if info and info.status_code == 200:
```
I hit this with scrolling wheel.
```
2017-03-30 03:31:22 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:22 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:22 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:22 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:22 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:23 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:23 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:23 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:23 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
2017-03-30 03:31:23 INFO Module `github _anon_module_0`: '----- GITHUB REQUEST -----'
```
Flaw?

**Comment by tobes** [comment]
> Flaw?

probably but github has a on_click() so would not be affected by this suggestion.  We could be a bit more clever about preventing the refresh.

**Comment by lasers** [comment]
Experiment with `deadbeef` module...
```
 diff --git a/py3status/modules/deadbeef.py b/py3status/modules/deadbeef.py
 index d70ca42d..028957b0 100644
 --- a/py3status/modules/deadbeef.py
 +++ b/py3status/modules/deadbeef.py
 @@ -93,6 +93,8 @@ class Py3status:

          deadbeef = deadbeef.strip().strip('-').strip(':')

 +        self.py3.log("----- DEADBEEF REQUEST -----")
 +
          return {
              'color': color,
              'cached_until': self.py3.time_in(self.cache_timeout),
```
I hit this hard with one nice scrolling wheel.
```
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method current_track returned {'full_text': '', 'color': '#ef2929', 'instance': '_anon_module_2', 'name': 'mpd_status'}
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
2017-03-30 08:38:37 INFO Module `deadbeef _anon_module_1`: '----- DEADBEEF REQUEST -----'
2017-03-30 08:38:37 INFO method deadbeef returned {'color': '#ef2929', 'full_text': 'DB', 'instance': '_anon_module_1', 'name': 'deadbeef'}
```

**Comment by tobes** [comment]
@lasers with `deadbeef` you look like you are not using the version in master.

**Comment by tobes** [comment]
What would be nice is to find this flaw in another module without an on_click that I can run.

**Comment by lasers** [comment]
Way ahead of you, buddy.
```
 diff --git a/py3status/modules/static_string.py b/py3status/modules/static_string.py
 index 3cdf273f..571c34b3 100644
 --- a/py3status/modules/static_string.py
 +++ b/py3status/modules/static_string.py
 @@ -13,11 +13,12 @@ class Py3status:
      """
      """
      # available configuration parameters
 -    format = ''
 +    format = '@@@@@ NEW STATIC UPDATE @@@@@'

      def static_string(self):
 +        self.py3.log("----- STATIC REQUEST -----")
          response = {
 -            'cached_until': self.py3.CACHE_FOREVER,
 +            'cached_until': self.py3.time_in(999999999999),
              'full_text': self.py3.safe_format(self.format),
          }
          return response
```

Out of `frame` this time. I don't see the log preventing anything?

```
2017-03-31 07:06:15 INFO Module `static_string`: '----- STATIC REQUEST -----'
2017-03-31 07:06:15 INFO method static_string returned {'full_text': '@@@@@ NEW STATIC UPDATE @@@@@', 'instance': '', 'name': 'static_string'}
2017-03-31 07:06:15 INFO received event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO trying to dispatch event to module "static_string"
2017-03-31 07:06:15 INFO dispatching event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO refresh py3status module static_string
2017-03-31 07:06:15 INFO clearing cache for method static_string
2017-03-31 07:06:15 INFO Module `static_string`: '----- STATIC REQUEST -----'
2017-03-31 07:06:15 INFO method static_string returned {'full_text': '@@@@@ NEW STATIC UPDATE @@@@@', 'instance': '', 'name': 'static_string'}
2017-03-31 07:06:15 INFO received event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO trying to dispatch event to module "static_string"
2017-03-31 07:06:15 INFO dispatching event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO refresh py3status module static_string
2017-03-31 07:06:15 INFO clearing cache for method static_string
2017-03-31 07:06:15 INFO Module `static_string`: '----- STATIC REQUEST -----'
2017-03-31 07:06:15 INFO method static_string returned {'full_text': '@@@@@ NEW STATIC UPDATE @@@@@', 'instance': '', 'name': 'static_string'}
2017-03-31 07:06:15 INFO received event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO trying to dispatch event to module "static_string"
2017-03-31 07:06:15 INFO dispatching event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO refresh py3status module static_string
2017-03-31 07:06:15 INFO clearing cache for method static_string
2017-03-31 07:06:15 INFO Module `static_string`: '----- STATIC REQUEST -----'
2017-03-31 07:06:15 INFO method static_string returned {'full_text': '@@@@@ NEW STATIC UPDATE @@@@@', 'instance': '', 'name': 'static_string'}
2017-03-31 07:06:15 INFO received event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO trying to dispatch event to module "static_string"
2017-03-31 07:06:15 INFO dispatching event {'name': 'static_string', 'instance': '', 'button': 4, 'x': 859, 'y': 1184}
2017-03-31 07:06:15 INFO refresh py3status module static_string
2017-03-31 07:06:15 INFO clearing cache for method static_string
2017-03-31 07:06:15 INFO Module `static_string`: '----- STATIC REQUEST -----'
2017-03-31 07:06:15 INFO method static_string returned {'full_text': '@@@@@ NEW STATIC UPDATE @@@@@', 'instance': '', 'name': 'static_string'}
```
Flaw?

**Comment by tobes** [comment]
I think we can close this now as the main issue was fixed in #866
